enhancements/network/allow-mtu-changes.md

```diff
@@ -0,0 +1,207 @@
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Customers may need to change the MTU post-installation. However these changes
+aren't trivial and may cause downtime, hence the CNO currently forbids them.
+
+We propose a procedure that will be launched on demand. These procedure will
+run pods on every node of the cluster and make the necessary changes in an
+ordered and coordinated manner with a service disruption within the least
+possible time and hopefully within TCP timeout (best effort).
+
+## Motivation
+
+While cluster administrators usually set the MTU correctly during the
+installation sometimes they need to change them afterwards for reasons such as
+changes in the underlay or because they were set incorrectly at install time.
+
+### Goals
+
+* Allow to change MTU post install on OVN Kubernetes.
+
+### Non goals
+
+* Change the MTU without service disruption.
+
+* Other safe or unsafe configuration changes.
+
+## Proposal
+
+The CNO monitors changes on the operator configuration. When it detects a MTU
+change:
+1. Set the `clusteroperator/network` conditions:
+   - Progressing: true
+   - Upgradeable: false
+2. Deploy pods on every node with `restartPolicy: Never` which are responsible
+   for validating the preconditions. If the preconditions are met the pod will
+   exit with code 0. Some of the preconditions that we will check are:
+   - For MTU changes, the new MTU is valid to apply on that node.
+3. Once all the previous pods are finished, deploy other set of pods on every
+   node that will handle the actual change of the MTU. Wait for them to be
+   ready and running.
+4. Cordon every node, we don't want pods created from this point on.
+5. Ensure via machine-config-operator, that upon reboot, configure-ovs will
+   set the br-ex and its physical port MTUs to the new MTU value.
+6. Ensure that the configmap ovnkube-config is synchronized with the new MTU
+   value.
+7. Set the new MTU value to the applied cluster configuration. At this point
+   CNO is commited to the MTU change.
+8. If any step was unsuccesful, the CNO will set the `clusteroperator/network`
+   conditions to:
+   - Progressing: false
+   - Degraded: true
+   Update the operator configuration status with a description of the problem.
+   At this point the process is interrupted and we require manual intervention.
+9. Wait for pods of step 3 to complete.
+10. Force a rollout of the ovnkube-master and ovnkube-node daemonset sequentially
+    and in appropriate order. This will ensure ovn-kubernetes uses the new MTU
+    value as well.
+11. If any of the previous steps failed, reboot the node, wait for the kubelet to
+    be reporting as Ready again and uncordon it. Otherwise just uncordon the node.
+12. Upon completion, set conditions to:
+    - Progressing: false
+    - Degraded: false
+
+The steps to change the MTU performed by pods of previous step 3 are:
```

Comment:
should this pod detect if there are MTU problems after migration is complete, and post an event or something to indicate if it was successful or not?

Reply:
Usually when you do a new deployment you ran some verification to check that the deployment has been done correctly and that cluster is healthy. If this exists for openshift we can suggest in documentation to run it again. Otherwise, it would probably be better to have a different set of pods with aliveness probe or the like rather than adding to these specific pods.

Reply:
we have the network check target pods, but I was thinking something specifically scoped to the MTU change to give the user a signal that the MTU update worked as part of the MTU update process itself. Like the pods that you launch for doing the MTU upgrade exit successfully and log some message like MTU upgrade complete, or if they check network connectivity and something is now broken, they either crash or post an event to their pod saying MTU upgrade problem. If you think it's not necessary then that's fine to ignore.

Reply:
I would probably then use the network check target pods and enhance that for any specific MTU verification we think we need to do. Do you know where I can check them out?

These MTU change pods only change the MTU of pods, which is an operation for which we should know definitively if it succeeded or not, and is only one step of a 3 step process which also includes changing the host sdn interfaces MTU and the host external interfaces MTU, so I feel that a verification of the MTU in these pods could be out of place.